### PR TITLE
feat: respondent journey (FPLA-2778/9)

### PR DESCRIPTION
**Pull request by joez-f**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/FPLA-2778
https://tools.hmcts.net/jira/browse/FPLA-2779

### Change description ###

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[X] No
```

### CODE REVIEW

File: ccd-definition/CaseEventToComplexTypes/enterRespondents.json [hunk 1]
```diff
@@ -0,0 +1,414 @@
...
+    "DisplayContext": "OPTIONAL"
+  },
+  {
+    "LiveFrom": "01/01/2017",
+    "ID": "EnterRespondents",
+    "CaseEventID": "enterRespondents",
+    "CaseFieldID": "respondents1",
+    "ListElementCode": "solicitor.email",
+    "FieldDisplayOrder": 3,
+    "DisplayContext": "OPTIONAL"
```
Comment (toireasam) [inline ccd-definition/CaseEventToComplexTypes/enterRespondents.json:227]: ticket says email should be mandatory, has this changed?

File: ccd-definition/CaseEventToComplexTypes/enterRespondents.json [hunk 2]
```diff
@@ -0,0 +1,414 @@
...
+    "FieldDisplayOrder": 4,
+    "DisplayContext": "OPTIONAL"
+  },
+  {
+    "LiveFrom": "01/01/2017",
+    "ID": "EnterRespondents",
+    "CaseEventID": "enterRespondents",
+    "CaseFieldID": "respondents1",
+    "ListElementCode": "solicitor.organisation.OrganisationName",
+    "DisplayContext": "OPTIONAL"
```
Comment (toireasam) [inline ccd-definition/CaseEventToComplexTypes/enterRespondents.json:244]: minor - fieldDisplayOrder seems to be missing from a number of fields below

File: ccd-definition/ComplexTypes/CareSupervision/1_UnregisteredOrganisation.json [hunk 3]
```diff
@@ -0,0 +1,34 @@
+[
+  {
+    "LiveFrom": "01/01/2017",
+    "ID": "UnregisteredOrganisation",
+    "ListElementCode": "signupRequestHeader",
+    "FieldType": "Label",
```
Comment (toireasam) [inline ccd-definition/ComplexTypes/CareSupervision/1_UnregisteredOrganisation.json:6]: minor - should this be signUpRequestHeader

File: ccd-definition/ComplexTypes/CareSupervision/RespondentNew-nonprod.json [hunk 4]
```diff
@@ -0,0 +1,19 @@
+[
+  {
+    "LiveFrom": "01/01/2017",
+    "ID": "RespondentNew",
+    "ListElementCode": "legalRepresentation",
+    "FieldType": "YesOrNo",
+    "ElementLabel": "Do they have legal representation?",
+    "SecurityClassification": "Public"
```
Comment (toireasam) [inline ccd-definition/ComplexTypes/CareSupervision/RespondentNew-nonprod.json:8]: In the prototype I see a label "Legal representation" should be before this radio button, unless design has changed

File: ccd-definition/CaseEventToComplexTypes/enterRespondents.json [hunk 5]
```diff
@@ -0,0 +1,414 @@
...
+  },
+  {
+    "LiveFrom": "01/01/2017",
+    "ID": "EnterRespondents",
+    "CaseEventID": "enterRespondents",
+    "CaseFieldID": "respondents1",
+    "ListElementCode": "legalRepresentation",
+    "DisplayContext": "OPTIONAL"
+  },
```
Comment (toireasam) [inline ccd-definition/CaseEventToComplexTypes/enterRespondents.json:201]: In the prototype I see a label "Legal representation" should be before this radio button, unless design has changed

File: e2e/tests/localAuthoritySubmitsApplication_@slow_test.js [hunk 6]
```diff
@@ -231,25 +232,27 @@ Scenario('local authority enters respondents @create-case-with-mandatory-section
...
   I.seeInTab(['Respondents 1', 'Telephone', 'Telephone number'], respondents[0].telephone);
-  I.seeInTab(['Respondents 1', 'What is the respondent\'s relationship to the child or children in this case?'], 'mock reason');
+  I.seeInTab(['Respondents 1', 'What is the respondent\'s relationship to the child or children in this case?'], respondents[0].relationshipToChild);
   I.seeInTab(['Respondents 1', 'Do you need contact details hidden from other parties?'], 'No');
   I.seeInTab(['Respondents 1', 'Do you believe this person will have problems with litigation capacity (understanding what\'s happening in the case)?'], 'Yes');
   I.seeInTab(['Respondents 1', 'Give details, including assessment outcomes and referrals to health services'], 'mock reason');
+  I.seeInTab(['Respondents 1', 'Representative', 'Representative\'s first name'], respondents[0].solicitor.firstName);
+  I.seeInTab(['Respondents 1', 'Representative', 'Representative\'s last name'], respondents[0].solicitor.lastName);
+  I.seeInTab(['Respondents 1', 'Representative', 'Email address'], respondents[0].solicitor.email);
+  //I.seeInTab(['Respondents 1', 'Representative', 'Organisation'], 'Swansea City Council');
```
Comment (toireasam) [inline e2e/tests/localAuthoritySubmitsApplication_@slow_test.js:250]: why is this one commented out?

File: ccd-definition/CaseEventToComplexTypes/enterRespondents.json (hunk 1 above)
Comment (joez-f) [reply ccd-definition/CaseEventToComplexTypes/enterRespondents.json:227]: nothing in mandatory in event, mandatory upon case submission

File: ccd-definition/CaseEventToComplexTypes/enterRespondents.json (hunk 5 above)
Comment (joez-f) [reply ccd-definition/CaseEventToComplexTypes/enterRespondents.json:201]: good spot, will add label

File: e2e/tests/localAuthoritySubmitsApplication_@slow_test.js (hunk 6 above)
Comment (joez-f) [reply e2e/tests/localAuthoritySubmitsApplication_@slow_test.js:250]: not sure 🤔 fixing

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/validators/RespondentsChecker.java [hunk 7]
```diff
@@ -49,17 +49,18 @@ private static boolean isEmptyRespondent(Respondent respondent) {
...
-                respondentParty.getContactDetailsHidden(),
-                respondentParty.getLitigationIssues());
+            && isEmptyTelephone(respondentParty.getTelephoneNumber())
+            && isEmptyEmail(respondentParty.getEmail())
+            && allEmpty(
+            respondentParty.getFirstName(),
+            respondentParty.getLastName(),
+            respondentParty.getDateOfBirth(),
+            respondentParty.getGender(),
+            respondentParty.getPlaceOfBirth(),
```
Comment (toireasam) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/validators/RespondentsChecker.java:59]: place of birth can probably be removed from here now?

File: service/src/main/java/uk/gov/hmcts/reform/fpl/model/RespondentSolicitor.java [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+
+@Data
+@Builder(toBuilder = true)
+@Jacksonized
+@Valid
+public class RespondentSolicitor {
+    private String firstName;
+    private String lastName;
+    @NotNull(message = "Enter the respondent's solicitor's email address")
+    private String email;
```
Comment (toireasam) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/model/RespondentSolicitor.java:19]: is this working? i seemed to be able to continue through the respondent event without email address

File: service/src/main/java/uk/gov/hmcts/reform/fpl/model/Respondent.java [hunk 9]
```diff
@@ -32,6 +37,34 @@
...
+    @Valid
+    private RespondentSolicitor solicitor;
+
+    @JsonIgnore
+    private boolean organisationSpecifiedForRespondentSolicitor;
+
+    @JsonIgnore
+    @AssertTrue(message = "Add organisation details for the respondent solicitor")
+    public boolean isOrganisationSpecifiedForRespondentSolicitor() {
+        if (YES.getValue().equals(legalRepresentation)) {
```
Comment (toireasam) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/model/Respondent.java:51]: is this used? intelij telling me it isn't

Comment (toireasam) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/model/Respondent.java:51]: discard this comment

File: service/src/main/java/uk/gov/hmcts/reform/fpl/model/Respondent.java [hunk 10]
```diff
@@ -32,6 +37,34 @@
...
+            //User selected yes but did not enter any details
+            if (isEmpty(solicitor)) {
+                return false;
+            }
```
Comment (toireasam) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/model/Respondent.java:55]: might it be worth adding a test for this in RespondentTest.java?

File: ccd-definition/CaseEventToComplexTypes/enterRespondents.json (hunk 1 above)
Comment (toireasam) [reply ccd-definition/CaseEventToComplexTypes/enterRespondents.json:227]: ah yes. :)

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/validators/RespondentsCheckerTest.java [hunk 11]
```diff
@@ -42,34 +46,121 @@ void shouldReturnErrorWhenNoRespondentsSpecified(List<Element<Respondent>> respo
...
+                .build())
+            .legalRepresentation(YES.getValue())
+            .solicitor(RespondentSolicitor.builder()
+                .unregisteredOrganisation(UnregisteredOrganisation.builder().name("Unregistered Org Name").build())
+                .firstName("Steve")
+                .email("steve@steve.com")
+                .build())
+            .build();
+        final CaseData caseData = CaseData.builder()
+            .respondents1(ElementUtils.wrapElements(respondent))
```
Comment (toireasam) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/service/validators/RespondentsCheckerTest.java:162]: might be good to add a test that no validation errors appear when legalRepresentation = NO

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/validators/RespondentsChecker.java (hunk 7 above)
Comment (joez-f) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/validators/RespondentsChecker.java:59]: yes 👍

File: service/src/main/java/uk/gov/hmcts/reform/fpl/model/Respondent.java (hunk 10 above)
Comment (joez-f) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/model/Respondent.java:58]: added tests for this logic

File: e2e/tests/localAuthoritySubmitsApplication_@slow_test.js (hunk 6 above)
Comment (joez-f) [reply e2e/tests/localAuthoritySubmitsApplication_@slow_test.js:250]: not trivial to fix as complex type for organisation is special and has different xpath in tab compared to normal field, maybe can be fixed in future tab story, as it is not a blocker here

File: ccd-definition/CaseEventToComplexTypes/enterRespondents.json (hunk 2 above)
Comment (joez-f) [reply ccd-definition/CaseEventToComplexTypes/enterRespondents.json:251]: hasn't been specified for some of the nested complex types, not sure what best approach is (whether or not it should be explicitly specified) - similar to existing definitions e.g enterChildren

Comment (toireasam) [reply ccd-definition/CaseEventToComplexTypes/enterRespondents.json:251]: sure, it's minor so fine

File: e2e/tests/localAuthoritySubmitsApplication_@slow_test.js (hunk 6 above)
Comment (toireasam) [reply e2e/tests/localAuthoritySubmitsApplication_@slow_test.js:250]: happy with this

File: service/src/test/java/uk/gov/hmcts/reform/fpl/model/RespondentTest.java [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+        void shouldReturnTrueWhenLegalRepresentationNeededAndRegisteredOrganisationSelected() {
+            Respondent respondent = Respondent.builder()
+                .legalRepresentation(YES.getValue())
+                .solicitor(RespondentSolicitor.builder()
+                    .firstName("Paul")
+                    .organisation(Organisation.builder().organisationID("test ID").build())
+                    .build())
+                .build();
+
+            assertTrue(respondent.isOrganisationSpecifiedForRespondentSolicitor());
```
Comment (toireasam) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/model/RespondentTest.java:55]: renaming to hasRequiredSolicitorOrganisationDetails() to make clearer

**Review by lewisbirks — DISMISSED** [review]

**Review by nraghu32 — DISMISSED** [review]

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/RespondentController.java [hunk 13]
```diff
@@ -86,6 +87,9 @@ public AboutToStartOrSubmitCallbackResponse handleAboutToSubmit(@RequestBody Cal
             caseData.getAllRespondents(), caseDataBefore.getAllRespondents()
         ));
 
+        caseData = getCaseData(caseDetails);
```
Comment (lewisbirks) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/RespondentController.java:90]: Don't do this, capture the previous assignment to the map as a variable and then remove from there. i.e.
```java
List<Element<Respondent>> respondents = respondentService.persistRepresentativesRelationship(...);
respondents = respondentService.removeHiddenFields(respondents);
caseDetails.getData().put(RESPONDENTS_KEY, respondents);
```

Comment (joez-f) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/RespondentController.java]: `        List<Element<Respondent>> respondents = respondentService.persistRepresentativesRelationship(
            caseData.getAllRespondents(), caseDataBefore.getAllRespondents());

        caseDetails.getData().put(RESPONDENTS_KEY, respondentService.removeHiddenFields(respondents));`
        did that

File: e2e/tests/localAuthoritySubmitsApplication_@slow_test.js [hunk 14]
```diff
@@ -231,29 +241,41 @@ Scenario('local authority enters respondents @create-case-with-mandatory-section
...
+  I.seeInTab(['Respondents 3', 'Party', 'First name'], respondents[2].firstName);
+  I.seeInTab(['Respondents 3', 'Party', 'Last name'], respondents[2].lastName);
+  I.seeInTab(['Respondents 3', 'Party', 'Date of birth'], '4 Apr 1978');
+  I.seeInTab(['Respondents 3', 'Party', 'Gender'], respondents[2].gender);
+  I.seeInTab(['Respondents 3', 'What is the respondent\'s relationship to the child or children in this case?'], respondents[2].relationshipToChild);
+  I.seeInTab(['Respondents 3', 'Do you need contact details hidden from other parties?'], 'No');
+  I.seeInTab(['Respondents 3', 'Do you believe this person will have problems with litigation capacity (understanding what\'s happening in the case)?'], 'No');
+  I.seeInTab(['Respondents 3', 'Representative', 'Representative\'s first name'], respondents[2].solicitor.firstName);
+  I.seeInTab(['Respondents 3', 'Representative', 'Representative\'s last name'], respondents[2].solicitor.lastName);
+  I.seeInTab(['Respondents 3', 'Representative', 'Email address'], respondents[2].solicitor.email);
```
Comment (Rob-Doherty) [inline e2e/tests/localAuthoritySubmitsApplication_@slow_test.js:277]: @joez-f would it be worth checking the additional unregistered details here too, something like (untested):
```suggestion
  I.seeInTab(['Respondents 3', 'Representative', 'Email address'], respondents[2].solicitor.email);
  I.seeInTab(['Respondents 3', 'Representative', 'Organisation (unregistered)', 'Organisation name'], respondents[2].solicitor.unregisteredOrganisation.name);
  I.seeInTab(['Respondents 3', 'Representative', 'Organisation (unregistered)', 'Address', 'Building and Street'], respondents[2].solicitor.unregisteredOrganisation.address.buildingAndStreet.lineOne);
  I.seeInTab(['Respondents 3', 'Representative', 'Organisation (unregistered)', 'Address', 'Address Line 2'], respondents[2].solicitor.unregisteredOrganisation.address.buildingAndStreet.lineTwo);
  I.seeInTab(['Respondents 3', 'Representative', 'Organisation (unregistered)', 'Address', 'Address Line 3'], respondents[2].solicitor.unregisteredOrganisation.address.buildingAndStreet.lineThree);
  I.seeInTab(['Respondents 3', 'Representative', 'Organisation (unregistered)', 'Address', 'Town or City'], respondents[2].solicitor.unregisteredOrganisation.address.town);
  I.seeInTab(['Respondents 3', 'Representative', 'Organisation (unregistered)', 'Address', 'Postcode/Zipcode'], respondents[2].solicitor.unregisteredOrganisation.address.postcode);
  I.seeInTab(['Respondents 3', 'Representative', 'Organisation (unregistered)', 'Address', 'Country'], respondents[2].solicitor.unregisteredOrganisation.address.country);
```

Comment (Rob-Doherty) [reply e2e/tests/localAuthoritySubmitsApplication_@slow_test.js:277]: These exact lines are included in a followup PR, so no point adding here.

Comment (joez-f) [reply e2e/tests/localAuthoritySubmitsApplication_@slow_test.js:277]: will add to follow up PR, currently it is missing, follow up only checks managing org not unregistered org

**Review by Rob-Doherty — COMMENTED** [review]
Have reviewed the auto test coverage and it looks good. Been able to get failures when expected. 👍

**Review by toireasam — APPROVED** [review]
LGTM

**Pull request merged**